### PR TITLE
Added generic filter_alleles, annotate_alleles.

### PR DESCRIPTION
**Pull request by cseed**
Also nuked filter_altered_genotypes which I don't think anyone was using, couldn't possibly have been the intended implementation, and can now be done generically.

I copied the existing annotate_alleles interface but, in light of generics, I don't like it anymore. I will propose an alternative on the dev forum.

### CODE REVIEW

File: src/main/scala/is/hail/variant/VariantDataset.scala [hunk 1]
```diff
@@ -309,7 +329,54 @@ g = let newgt = ${ filterGT("gtIndex(oldToNew[gtj(g.gt)], oldToNew[gtk(g.gt)])")
...
+    } else {
+      if (!vsm.genotypeSignature.isOfType(Genotype.htsGenotypeType))
+        fatal(s"filter_alleles: genotype_schema must be TGenotype or the HTS genotype schema, found: ${ vsm.genotypeSignature }")
+
+      if (subset) {
+        """
+g = let newpl = if (isDefined(g.PL))
+        let unnorm = range(newV.nGenotypes).map(newi =>
+            let oldi = gtIndex(newToOld[gtj(newi)], newToOld[gtk(newi)])
+             in g.PL[oldi]) and
```
Comment (jbloom22) [inline src/main/scala/is/hail/variant/VariantDataset.scala]: These expressions look right. The indentation is strange in places like here.

File: python/hail/dataset.py [hunk 2]
```diff
@@ -1661,24 +1732,166 @@ def write(self, output, overwrite=False):
...
+        ...           else
+        ...             NA: Array[Int] and
+        ...         newgq = gqFromPL(newpl)
+        ...      in { GT: Call(newgt), AD: newad, DP: newdp, GQ: newgq, PL: newpl }''',
+        ...     keep=False)
+
+        **Notes**
+
+        ``expr`` is a boolean condition indicating if the given allele
+        should be kept or removed.  ``aIndex`` (allele index)
```
[the remark this answers is not in the document]
Comment (cseed) [reply python/hail/dataset.py]: That's not true, it depends on the value of keep.

File: src/main/scala/is/hail/variant/VariantDataset.scala (hunk 1 above)
Comment (cseed) [reply src/main/scala/is/hail/variant/VariantDataset.scala]: I'm going to leave it.  We don't really have an indentation convention and it is all going to become python soon anyway.

File: python/hail/dataset.py [hunk 3]
```diff
@@ -199,15 +200,15 @@ def annotate_alleles_expr_generic(self, split_variant_expr, split_genotype_expr,
...
+        - ``aIndex``: the index of the input allele corresponding to the output variant
         - ``wasSplit``: True if the original variant was multiallelic
         - ``s``: the sample
         - ``sa``: the sample annotations
         - ``g``: the old genotype annotations
 
-        ``variant_expr`` is an annotation expression that the original
-        variant annotations per-allele.  It updates ``va`` and is
-        evaluated in the following namespace:
+        ``variant_expr`` is an annotation expression that updated the
```
Comment (jbloom22) [inline python/hail/dataset.py]: updated => updates

**Pull request merged**